src/Xamarin.Android.Build.Tasks/Tasks/BuildApk.cs

```diff
@@ -133,25 +169,28 @@ void ExecuteWithAbi (string [] supportedAbis, string apkInputPath, string apkOut
 				AddRuntimeLibraries (apk, supportedAbis);
 				apk.Flush();
 				AddNativeLibraries (files, supportedAbis);
-				apk.Flush();
 				AddAdditionalNativeLibraries (files, supportedAbis);
-				apk.Flush();
 
 				if (TypeMappings != null) {
 					foreach (ITaskItem typemap in TypeMappings) {
-						apk.Archive.AddFile (typemap.ItemSpec, RootPath + Path.GetFileName(typemap.ItemSpec), compressionMethod: UncompressedMethod);
+						AddFileToArchiveIfNewer (apk, typemap.ItemSpec, RootPath + Path.GetFileName(typemap.ItemSpec), compressionMethod: UncompressedMethod);
 					}
 				}
 
 				int count = 0;
 				foreach (var file in files) {
-					var item = Path.Combine (file.Item2, Path.GetFileName (file.Item1))
+					var item = Path.Combine (file.archivePath, Path.GetFileName (file.filePath))
 						.Replace (Path.DirectorySeparatorChar, '/');
+					existingEntries.Remove (file.filePath);
 					if (apk.Archive.ContainsEntry (item)) {
-						Log.LogCodedWarning ("XA4301", file.Item1, 0, "Apk already contains the item {0}; ignoring.", item);
-						continue;
+						DateTime lastWrite = File.GetLastWriteTimeUtc (file.filePath);
+						if (lastWrite <= apk.Archive.ReadEntry (item).ModificationTime || !refresh) {
+							Log.LogCodedWarning ("XA4301", file.filePath, 0, "Apk already contains the item {0}; ignoring.", item);
```

Comment:
I need to figure out what to do with this Warning now.. its causing a few tests to break.